sscanss/editor/json_attributes.py

```diff
@@ -0,0 +1,858 @@
+
+        label = QtWidgets.QLabel(self.title)
+        label.setMaximumWidth(120)
+        label.setMinimumWidth(120)
+        widget.layout.addWidget(label)
+        edit_widget = self.value.createEditWidget(self.title)
+        widget.layout.addWidget(edit_widget)
+
+        checkbox = QtWidgets.QCheckBox("Add")
+        checkbox.setMinimumWidth(50)
+        checkbox.setMaximumWidth(50)
+        if not self.mandatory:
+            checkbox.setChecked(self.turned_on)
+            checkbox.stateChanged.connect(self.setTurnedOn)
+        else:
+            size_policy = checkbox.sizePolicy()
+            size_policy.setRetainSizeWhenHidden(True)
+            checkbox.setSizePolicy(size_policy)
+            checkbox.hide()
+
+        widget.layout.addWidget(checkbox)
+
+        return widget
+
+    def connectParent(self, parent):
+        """Connects the parent for the value inside
+        :param parent: the parent to connect to the value
+        :type parent: JsonValue
+        """
+        self.value.connectParent(parent)
+        self.been_set.connect(parent.been_set.emit)
+
+    @property
+    def json_value(self):
+        """Returns the json value of itself"""
+        return self.value.json_value
+
+    @json_value.setter
+    def json_value(self, value):
+        """Sets the value from the json representation"""
+        self.value.json_value = value
+
+
+class JsonValue(QtCore.QObject):
+    """The parent class of all the nodes in the tree
+    Contains the methods which should be overridden by the child classes
+    :param initial_value: the initial value of the object, if left empty will be overwritten by the default value
+    :type initial_value: object
+    """
+    been_set = QtCore.pyqtSignal(object)
+    default_value = None
+
+    def __init__(self, initial_value=None):
+        super().__init__()
+        self.parent = False
+        if initial_value:
+            self.value = initial_value
+        else:
+            self.value = self.default_value
+
+    def createEditWidget(self, title=''):
+        """Creates the widget which should be used in the UI to edit the data.
+        It is linked with the object using events
+        :param title: title the which should be used inside the widget if it is needed
+        :type title: str
+        """
+        return QtWidgets.QWidget()
+
+    def defaultCopy(self):
+        """Creates a copy of the attribute, it is needed to be able to create an object,
+        following the same schema as another object
+        :return: the copy of self, following the same schema
+        :rtype: JsonValue
+        """
+        return type(self)()
+
+    def connectParent(self, parent):
+        """Connects self to the parent node in the tree needed to be able to backtrack the tree and send the
+        events up the tree
+        :param parent: the parent of the node
+        :type parent: JsonValue
+        """
+        self.been_set.connect(parent.been_set.emit)
```

Comment:
Both the attribute and the value emit the same signal. Is this intended?

Reply:
It is probably not the best design decision to have same signal name but it is intended as attribute handles the checkbox and the ability to switch on/off optional attributes, so JSON is modified not only when value changes but also when the checkbox is switched.